src/java.base/share/classes/jdk/internal/reflect/MethodHandleAccessorFactory.java

```diff
@@ -347,36 +348,34 @@ static void ensureClassInitialized(Class<?> defc) {
      * Native accessor, i.e. VM reflection implementation, is used if one of
      * the following conditions is met:
      * 1. during VM early startup before method handle support is fully initialized
-     * 2. a Java native method
-     * 3. -Djdk.reflect.useNativeAccessorOnly=true is set
+     * 2. -Djdk.reflect.useNativeAccessorOnly=true is set
+     * 3. a signature polymorphic method
      * 4. the member takes a variable number of arguments and the last parameter
      *    is not an array (see details below)
      * 5. the member's method type has an arity >= 255
      *
+     * Conditions 3-5 are due to the restrictions of method handles.
      * Otherwise, direct invocation of method handles is used.
      */
     private static boolean useNativeAccessor(Executable member) {
         if (!VM.isJavaLangInvokeInited())
             return true;
 
-        if (Modifier.isNative(member.getModifiers()))
+        if (ReflectionFactory.useNativeAccessorOnly())  // for testing only
             return true;
 
-        if (ReflectionFactory.useNativeAccessorOnly())  // for testing only
+        // java.lang.invoke cannot find the underlying native stubs of signature
+        // polymorphic methods that core reflection must invoke.
+        // Fall back to use the native implementation instead.
+        if (member instanceof Method method && isSignaturePolymorphicMethod(method))
             return true;
 
-        // MethodHandle::withVarargs on a member with varargs modifier bit set
-        // verifies that the last parameter of the member must be an array type.
-        // The JVMS does not require the last parameter descriptor of the method descriptor
-        // is an array type if the ACC_VARARGS flag is set in the access_flags item.
-        // Hence the reflection implementation does not check the last parameter type
-        // if ACC_VARARGS flag is set.  Workaround this by invoking through
-        // the native accessor.
-        int paramCount = member.getParameterCount();
-        if (member.isVarArgs() &&
-                (paramCount == 0 || !(member.getParameterTypes()[paramCount-1].isArray()))) {
+        // java.lang.invoke fails to create MH for bad ACC_VARARGS methods with no
+        // trailing array,  but core reflection ignores ACC_VARARGS flag like the JVM does.
+        // Fall back to use the native implementation instead.
+        if (isInvalidVarArgs(member))
```

Comment:
I prefer to keep the check inlined here rather than an utility method as it's explicit what to check.